### PR TITLE
TIME - Angela

### PR DESCRIPTION
**Pull request by angethuy**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? |  I spent a lot of time considering making blocks a child class of Reservation but I didn't think the relationship warranted managing inheritance when the difference between a Reservation and a Block was just 1) more rooms and guests and 2) an optional discount rate.
What was a design decision you made that changed over time over the project? |  I chose to make "occupancy" a special variable - it tracks rooms belonging to a reservation (one to [1-5] relationship) as an array of hashes with room id, guest name. This was a pain to manage. I'm not sure I would dig myself into the same rabbit hole in the future.
What was a concept you gained clarity on, or a learning that you'd like to share? | Overlapping dates are much simpler than they seem!!! If we can trust that a date range implies contiguous dates, then to test overlap, we need to check if 1) a proposed start date falls inside of the range EXCLUDING end date, or or 2) the proposed end date falls inside of the range EXCLUDING start date. We only need to check for two cases.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | Testing the creation of a new Reservation is a nominal test because it directly ties to a user story.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | I had an edge case to verify that an ArgumentError is raised if I try to make a reservation where the end date occurs before the start date. It's an edge case because it's a wild and whacky situation that should not occur but just in case it does... 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I didn't spend enough time writing pseudocode and will likely dedicate more time to that step for my next project. I'm still having trouble writing tests before my code but my process for writing code and tests occurs more in tandem now compared to my habits from before. Using guard and coverage have helped a lot.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -48,3 +48,11 @@ build-iPhoneSimulator/
...
 # unless supporting rvm < 1.11.0 or doing something fancy, ignore this:
 .rvmrc
+
+#simplecov
+
+coverage
+
+#system
+
+.DS_store
```
Comment (kaidamasaki) [inline .gitignore:58]: You can ignore `.DS_Store` everywhere by creating a global gitignore file.

https://help.github.com/en/github/using-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+require 'date'
+
+module Hotel
+  class DateRange
+    attr_reader :check_in, :check_out, :nights_spent
+
+    def initialize(check_in, check_out)
+      raise ArgumentError, 'Invalid date range given.' unless check_out > check_in
```
Comment (kaidamasaki) [inline lib/date_range.rb:8]: It's helpful to include the invalid arguments when you raise an `ArgumentError`.

```suggestion
      raise ArgumentError, "Invalid date range given. check_in: #{check_in} check_out: #{check_out}" unless check_out > check_in
```

File: lib/date_range.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      @check_in = check_in
+      @check_out = check_out
+      @nights_spent = (@check_out - @check_in)
```
Comment (kaidamasaki) [inline lib/date_range.rb:11]: Parenthesis here are unnecessary:

```suggestion
      @nights_spent = @check_out - @check_in
```

File: lib/date_range.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    def initialize(check_in, check_out)
+      raise ArgumentError, 'Invalid date range given.' unless check_out > check_in
+      @check_in = check_in
+      @check_out = check_out
+      @nights_spent = (@check_out - @check_in) 
+    end
+
+    def to_id
+      return (@check_in.year.to_s + @check_in.mon.to_s + @check_in.mday.to_s)
```
Comment (kaidamasaki) [inline lib/date_range.rb:15]: You don't need parenthesis around return values:

```suggestion
      return @check_in.year.to_s + @check_in.mon.to_s + @check_in.mday.to_s
```

File: lib/date_range.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    end
+
+    def overlap? (other_check_in, other_check_out)
+      return ((@check_in..@check_out).cover? (other_check_in)) || ((@check_in-1...@check_out).cover? (other_check_out))
```
Comment (kaidamasaki) [inline lib/date_range.rb:19]: Clever use of `cover?`!

File: lib/reservation.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    attr_reader :id, :dates, :occupancy 
+
+    def initialize(type, dates, occupancy)
+      @id = type[0] + dates.to_id 
+      @dates = dates
+      @occupancy = occupancy
+    end
+
+    def total_price 
+        return @occupancy.sum { |occupancy| occupancy[:room].price * @dates.nights_spent }
```
Comment (kaidamasaki) [inline lib/reservation.rb:17]: Incorrect indentation:

```suggestion
      return @occupancy.sum { |occupancy| occupancy[:room].price * @dates.nights_spent }
```

File: lib/reservation_manager.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    private 
+
+    def generate_rooms(room_count)
+      rooms = []
+      room_count.times do |num|
+        rooms << Room.new(num+1)
+      end
+      return rooms
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:63]: Good use of a private method!  

You can simplify this a little using a range and `map` though:
```suggestion
      return (1..room_count).map do |num|
        Room.new(num)
      end
```

File: lib/reservation_manager.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    # Parameters: 
+    # -type: symbol (:SINGLE, :BLOCK) that indicates type of reservation
+    # -range: DateRange object 
+    # -occupancy: Hash or String
+    # Returns: newly created Reservation object
+    def create_reservation(type, range, occupancy)
+      raise ArgumentError.new("Invalid reservation type (given #{type}, must be :SINGLE or :BLOCK)") unless 
+        type == :SINGLE || type == :BLOCK
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:22]: Great use of a guard clause with a good message!

File: lib/reservation_manager.rb [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+      occupancy = [{:room => @rooms.sample, :guest => occupancy}] if occupancy.class == String
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:24]: If you're going to check a type I'd recommend using `kind_of?` instead of `class` since that way you allow subclasses too:

```suggestion
      occupancy = [{:room => @rooms.sample, :guest => occupancy}] if occupancy.kind_of? String
```

File: lib/reservation_manager.rb [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    def initialize(room_count)
+      @rooms = generate_rooms(room_count)
+      @reservations = []
+    end
+
+    # Parameters: 
+    # -type: symbol (:SINGLE, :BLOCK) that indicates type of reservation
+    # -range: DateRange object 
+    # -occupancy: Hash or String
+    # Returns: newly created Reservation object
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:19]: I really like that you documented your parameters but it would be more helpful if you mentioned more than just the type but also what they represent.

Really the only confusing bit here is occupancy.  Everything else is pretty self explanatory or well explained (like `type`).

File: lib/reservation_manager.rb [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+      new_reservation = Reservation.new(type, range, occupancy)
+      @reservations << new_reservation
+      return new_reservation
+    end
+
+    # Parameters: Date object 
+    # Returns: an Array of all Reservation instances that contain that date
+    def find_reservations_by_date(check_in, check_out)
+      by_date = @reservations.select { |reservation| reservation.dates.overlap?(check_in, check_out)}
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:34]: This works because of implicit return, but I don't think you meant to assign here (since you don't use the variable):

```suggestion
      return @reservations.select { |reservation| reservation.dates.overlap?(check_in, check_out)}
```

File: lib/reservation_manager.rb [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    def find_reservations_by_room(room)
+      by_room = @reservations.select { |reservation| reservation.occupancy.detect { |occupancy| occupancy[:room] == room }}
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:38]: See above.

```suggestion
      return @reservations.select { |reservation| reservation.occupancy.detect { |occupancy| occupancy[:room] == room }}
```

File: test/reservation_manager_test.rb [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      expect(@reservation_manager).must_be_kind_of Hotel::ReservationManager
+    end
+
+    it "initializes correct number of rooms" do
+      expect(@reservation_manager.rooms.length).must_equal 20
+    end
+
+    it "populates valid room objects" do
+      expect(@reservation_manager.rooms[1]).must_be_kind_of Hotel::Room
+    end
```
Comment (kaidamasaki) [inline test/reservation_manager_test.rb:19]: A better check here would be to loop through all of the rooms with individual assertions:

```suggestion
    it "populates valid room objects" do
      @reservation_manager.rooms.each do |room|
        expect(room).must_be_kind_of Hotel::Room
      end
    end
```

File: test/reservation_manager_test.rb [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+      reservation_manager = Hotel::ReservationManager.new(0)
+      expect(reservation_manager.rooms.length).must_equal 0
+
+    end
+  end
+
+  describe "create reservation" do
+    before do
+      @room = Hotel::Room.new(100)
+      @occupancy = [{:room => @room, :guest => "Picchu"}]
```
Comment (kaidamasaki) [inline test/reservation_manager_test.rb:35]: Is this like the pokemon? 

Pichu

**Comment by kaidamasaki** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | Somewhat and is making good progress.
Practices git with at least 15 small commits and meaningful commit messages | Please commit more.  I promise it will be helpful later.  :)

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | No tests on `overlaps?`.
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️ :tada: 100% :tada:

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | No edge case test.
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | No edge case test.
creates a block of rooms | ✔️
reserves a room from a block | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you organized your methods within your classes and made good use of Ruby's built-in methods.

I do see some room for improvement around testing edge cases.

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation |
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Here are a few little things that you can do to clean up your code even more.  😄
